src/engine/sparqlExpressions/StringExpressions.cpp

```diff
@@ -202,7 +210,21 @@ class SubstrImpl {
       lengthInt += startInt;
     }
 
-    const auto& str = s.value();
+    const auto& str = asStringViewUnsafe(s.value().getContent());
+    std::optional<std::variant<Iri, std::string>> descriptor;
+
+    if (s->isLiteral()) {
+      if (s->hasLanguageTag()) {
+        descriptor = std::string(asStringViewUnsafe(s->getLanguageTag()));
+      } else if (s->hasDatatype()) {
+        descriptor =
+            ad_utility::triple_component::Iri::fromIrirefWithoutBrackets(
+                asStringViewUnsafe(s->getDatatype()));
+      }
+    } else {
```

Comment:
It would be more efficient ( you get your own copy of the LiteralOrIri after all), to not create a new LiteralOrIri,
But to implement the `SubStr` directly inside the `Literal` class, s.t. you can reuse the allocation etc, this should make things somewhat faster. So a method `Literal::setSubtr(...)` would be the way to go.

Reply:
I have tried to implement the method setSubStr in Literal.cpp. However, I have to access the Literal object in StringExpressions.cpp with getLiteral() to execute setSubStr(). getLiteral() is a const function. I have now added a getLiteral() without const. Is that correct? Otherwise I would have to copy the object again.

Reply:
A non-const-getter for modifying is a good way to do this.